python/nano/src/bigdl/nano/tf/keras/customized_training_utils.py

```diff
@@ -169,46 +185,79 @@ def _train_func(target_path, *args):
-            actrual_args[i] = loss_object
-            continue
-        # deserialize others
-        if arg[0] == "others":
-            with open(arg[1], 'rb') as f:
-                actrual_args[i] = cloudpickle.load(f)
-                if callable(actrual_args[i]) and isinstance(actrual_args[i], nano_multiprocessing):
-                    actrual_args[i] = partial(actrual_args[i], mirrored_strategy=mirrored_strategy)
-
-    with open(target_path, 'rb') as f:
-        target_func = cloudpickle.load(f)
-
-    res = target_func(*actrual_args)
-
-    task_id = mirrored_strategy.cluster_resolver.task_id
-    # TODO: only task 0's model weight is stored
-    # could not understand why we need other task's weight
-    if task_id == 0:
-        path = os.path.join('trained_model_weights')
-        new_model.save_weights(path, overwrite=True)
+            # deserialize dataset
+            if arg[0] == "dataset":
+                try:
+                    from tensorflow.python.distribute.coordinator.values import \
+                        deserialize_dataset_from_graph
+                    original_dataset = deserialize_dataset_from_graph(arg[1], arg[2])
+                except ValueError:
+                    original_dataset = tf.data.Dataset.from_generator(arg[1],
+                                                                      output_signature=arg[2])
+                actrual_args[i] = mirrored_strategy.experimental_distribute_dataset(
+                    original_dataset)
+                continue
 
+            if arg[0] == "loss":
+                with open(arg[1], 'rb') as f:
+                    original_loss_object = cloudpickle.load(f)
+                    original_loss_object.reduction = tf.keras.losses.Reduction.NONE
+
+                def loss_object(*args, **kwargs):
+                    per_example_loss = original_loss_object(*args, **kwargs)
+                    if per_example_loss.shape == [] or per_example_loss.shape[0] == 0:
+                        size = mirrored_strategy.num_replicas_in_sync
+                        return tf.math.reduce_sum(per_example_loss) / size
+                    else:
+                        size = per_example_loss.shape[0] * mirrored_strategy.num_replicas_in_sync
+                        return tf.nn.compute_average_loss(per_example_loss, global_batch_size=size)
+                actrual_args[i] = loss_object
+                continue
+            # deserialize others
+            if arg[0] == "others":
+                with open(arg[1], 'rb') as f:
+                    actrual_args[i] = cloudpickle.load(f)
+                    if callable(actrual_args[i]) and \
+                       isinstance(actrual_args[i], nano_multiprocessing):
+                        actrual_args[i] = partial(actrual_args[i],
+                                                  mirrored_strategy=mirrored_strategy)
+
+        with open(target_path, 'rb') as f:
```

Comment:
@ACupofAir please have a look at this PR and add some security related change after its merging